src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+            // Look into the first byte to see if we are connecting to a 6.5 or earlier server.
+            int option = await tdsStream.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+
+            if (option == 0xaa)
+            {
+                // If the first byte is 0xAA, we are connecting to a 6.5 or earlier server, which
+                // is not supported.
+                throw SQL.InvalidSQLServerVersionUnknown();
+            }
+
+            byte[] payload = new byte[tdsStream.PacketDataLeft];
+
+            int payloadOffset = 0;
+            int offset = 0;
+
+            while (option != (byte)PreLoginOptions.LASTOPT)
+            {
+                switch (option)
+                {
+                    case (int)PreLoginOptions.VERSION:
+
+                        payloadOffset = payload[offset++] << 8 | payload[offset++];
+                        offset += PAYLOAD_LENGTH_SIZE_IN_BYTES; // Skip the payload length
+
+                        byte majorVersion = payload[payloadOffset];
+
+                        bool is2005OrLater = majorVersion >= 9;
+                        if (!is2005OrLater)
+                        {
+                            context.ConnectionContext.MarsCapable = false;  // If pre-2005, MARS not supported.
+                        }
+
+                        break;
+
+                    case (int)PreLoginOptions.ENCRYPT:
+                        if (context.IsTlsFirst)
+                        {
+                            // Can skip/ignore this option if we are doing TDS 8.
+                            // The internal encryption option is set to NOT_SUP in this case.
+                            // And it shouldn't be changed.
+                            offset += 4;
+                            break;
+                        }
+
+                        payloadOffset = payload[offset++] << 8 | payload[offset++];
+                        offset += 2; // Skip the payload length
+
+                        EncryptionOptions serverOption = (EncryptionOptions)payload[payloadOffset];
+
+                        // Any response other than NOT_SUP means the server supports encryption.
+                        context.ServerSupportsEncryption = serverOption != EncryptionOptions.NOT_SUP;
+
+                        // If server doesn't support encyrption, then we can't proceed, since encryption is always needed
+                        // for login.
+                        if (!context.ServerSupportsEncryption)
+                        {
+                            //    _physicalStateObj.AddError(new SqlError(TdsEnums.ENCRYPTION_NOT_SUPPORTED, (byte)0x00, TdsEnums.FATAL_ERROR_CLASS, _server, SQLMessage.EncryptionNotSupportedByServer(), "", 0));
+                            //    _physicalStateObj.Dispose();
+                            //    ThrowExceptionAndWarning(_physicalStateObj);
+                            context.ConnectionContext.Error = new Exception("Encryption not supported by server");
```

Comment:
Yea, we will reuse the existing Resource String. I didn't have a good point of view about Exception handling. I brainstormed this with @samsharma2700  who educated me about how the exception handling should work. 
Essentially all handlers will throw, and while throwing create a SqlError object based on the existing exception messages and then those will be wrapped in SqlException.

Reply:
while working on exceptions, we issue #26 needs to be fixed as well.

cc: @roji

Reply:
@JRahnama lets not mix the cancellation exception discussion into a prelogin PR 😀
The only decision we are taking here is that all exceptions will be thrown from IO instead of the old SNI style of saving the exception and retrieving it later.

Reply:
yeah, it just past my mind. I'll keep that for a related PR.